### PR TITLE
Make binary() aka String.t() in Elixir refinable

### PR DESCRIPTION
**Pull request by erszcz**
`String.t()` is the default string type in Elixir just as `string()` is the default in Erlang. However, when Erlang `string()` is a _charlist_, i.e. a list of integers, the Elixir default string type is actually a binary. This makes binaries ubiquitous in Elixir code, much more so than in Erlang. To make exhaustiveness checking useful and relevant in Elixir (for example in cases such as #391), it has to support the binary type.

Alas, this brings in some problems, the most obvious one being a broken test case. Therefore, until this is solved, I'm making this a draft.

### CODE REVIEW

**Comment by zuiderkwast** [comment]
I don't really know the exact meaning of `refinable/3`. @josefs added it, for exhaustiveness-checking, but I don't think it's used for type refinement.

For refinement, we can only partially refine a type if a pattern matches a subtype of the type (e.g. a singleton so that we can subtract it from the type (e.g. `x|y \ x ==> y`). However, a free variable or `_` matches everything and this should be possible to use regardless of whether a type has singleton subtypes or not.

What I'm trying to say is that if we can make binary() refinable, we can probably remove refinable alltogether.

(Additionally, we could maybe refine by the empty binary which is a singleton: `binary() \ <<>> ==> <<_:8, _:_*8>>` (i.e. nonempty-binary).)

**Comment by erszcz** [comment]
> I don't really know the exact meaning of refinable/3. @josefs added it, for exhaustiveness-checking, but I don't think it's used for type refinement.

It's not used for type refinement per se. It's used solely as a gatekeeper to exhaustiveness checking. We had a conversation about it in some other issue already. In short, "if a type is refinable we can check exhaustiveness."

> we could maybe refine by the empty binary which is a singleton: binary() \ <<>> ==> <<_:8, _:_*8>> (i.e. nonempty-binary).

Good idea :+1:

**Comment by erszcz** [comment]
> We had a conversation about it in some other issue.
https://github.com/josefs/Gradualizer/pull/347

**Comment by berbiche** [comment]
Related? https://github.com/josefs/Gradualizer/issues/339#issuecomment-895786061

**Comment by zuiderkwast** [comment]
> Related? [#339 (comment)](https://github.com/josefs/Gradualizer/issues/339#issuecomment-895786061)

Maybe. I don't remember anymore. :-)

@erszcz Do you think everything except any() should be refineable?

**Comment by erszcz** [comment]
> @erszcz Do you think everything except any() should be refineable?

I think so. From the opposite perspective, would we want any types to disable exhaustiveness checking? Ideally not, I think. Therefore everything except `any()` should be refinable. Whether it's feasible for everything except `any()` is another question, but I think we're already quite close.

**Comment by zuiderkwast** [comment]
Good! So if you keep any other types as non-refineable, you can add a comment about why? (missing implementation in other parts of the code, etc.)

**Comment by erszcz** [comment]
This is now ready for review. I've followed a similar pattern to that of checking empty/nonempty lists, i.e. we only check simple patterns and for complex ones exhaustiveness checking is disabled. In general making `binary()`/`String.t()` refinable should unblock exhaustiveness checking on complex types having it as a component.

`make gradualize` status is as follows - before:

```
$ wc -l gradualize.002f5f7.log
     307 gradualize.002f5f7.log
```

with this PR:

```
$ wc -l gradualize.3108bcd.log
     310 gradualize.3108bcd.log
```

The increase in self gradualization errors is caused by a new warning being emitted:

```
ebin/gradualizer_db.beam: Nonexhaustive patterns on line 608 at column 17
Example values which are not covered:
	match
```

This warning is now emitted since `re:run/3` present on the reported line returns complex terms containing binaries. This means we were not able to check its return value for exhaustiveness prior to this change and we are able to do so now.

**Comment by erszcz** [comment]
With the latest fix we're back to

```
$ wc -l gradualize.b185fb5.log
     307 gradualize.b185fb5.log
```

File: src/gradualizer_db.erl
Comment (zuiderkwast) [inline src/gradualizer_db.erl:613]: Debugging code? We already get a case clause error without this, so I don't think we should add this explicit code. Or is this just to silence a gradualizer error?

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:197]: `<<>>` is not a value of every bitstring type. The type might have a fixed or minimum length >= 0.

File: src/typechecker.erl
Comment (zuiderkwast) [inline src/typechecker.erl:3727]: The `8` here can be turned into a variable N to make this a bit more generic. We just need a guard `when N > 0`.

File: src/typechecker.erl
Comment (zuiderkwast) [inline src/typechecker.erl:4442]: A nice way to remember this is to add a known problem where we e.g. have a bound variable V in `<<..., V/binary>>`.

**Review by zuiderkwast — COMMENTED** [review]
Can you explain in which cases we can do exhaustiveness checking for bitstring types and patterns and in which cases we can't?

For example, the following is exhaustive but I don't think we can capture it, can we?

```Erlang
-spec f(<<_:1, _:_*0>>) -> ok.
f(<<0:1>>) -> ok;
f(<<1:1>>) -> ok.
```

Do we get a false positive for the above?

What about the `<<..., _/binary>>` and `<<..., _/bitstring>>` patterns? What are the guarantees?

File: src/gradualizer_db.erl
Comment (erszcz) [reply src/gradualizer_db.erl:613]: > is this just to silence a gradualizer error?

It is, exactly :+1: We don't get the warning for remaining/unhandled `re:run/3` return values anymore as we handle whatever else might be returned explicitly.

In general, I'm trying out approaches to deal with warnings that will reported by Gradualizer. A clause like this seems reasonable as it doesn't require extra options in the typechecker or pragmas in code and is also a reminder to the readers that the catch-all is intentional.

**Comment by erszcz** [comment]
> Can you explain in which cases we can do exhaustiveness checking for bitstring types and patterns and in which cases we can't?

This is probably a bit flaky currently. We only do exhaustiveness checking in case of patterns covering:
- empty binaries
- nonempty binaries, that is binaries with exactly two segments (i.e. `bin_element` tuples) and the `bytes`/`binary` specifier for the last segment; first segment size is not taken into consideration

It's based on how we do it for lists.

> For example, the following is exhaustive but I don't think we can capture it, can we?
> 
> ```erlang
> -spec f(<<_:1, _:_*0>>) -> ok.
> f(<<0:1>>) -> ok;
> f(<<1:1>>) -> ok.
> ```
> 
> Do we get a false positive for the above?

We don't as the patterns are single-segment, so they disable exhaustiveness checking:

```
$ cat t.erl
-module(t).

-spec f(<<_:1, _:_*0>>) -> ok.
f(<<0:1>>) -> ok;
f(<<1:1>>) -> ok.
$ ./bin/gradualizer --verbose t.erl
...
Exhaustiveness checking: false
Popping clauses controls
Task returned from function f/1 with []
```

> What about the `<<..., _/binary>>` and `<<..., _/bitstring>>` patterns? What are the guarantees?

`bitstring` disables exhaustiveness checking. `binary` doesn't, unless the ellipsis is a single segment, as mentioned above.

File: src/gradualizer_db.erl
Comment (zuiderkwast) [reply src/gradualizer_db.erl:613]: It effectively silences the warning, but it goes against good Erlang style by adding defensive code... Well well, it's partly because of re:run/3 returning different stuff depending on options.

**Comment by zuiderkwast** [comment]
So anything except a few cases disables exhaustiveness checking. That is, it is best effort and we don't try to be sound regarding exhaustiveness checking in genereal. It's good to be clear about that, since we want to be sound (gradually sound, the graduality property, whatever) in the main parts of the type system.

I repeat the wish for a page about exhaustiveness checking in the https://github.com/josefs/Gradualizer/wiki.

File: src/typechecker.erl
Comment (zuiderkwast) [inline src/typechecker.erl:4451]: This function doesn't check if a pattern is exhaustive (what the name indicates) but rather if the pattern can be used for exhaustiveness checking or if exhaustiveness checking should be disabled.

I think we should come up with a better name for this function. (We can negate the return values or change them if it fits the function name better.)

Comment (erszcz) [reply src/typechecker.erl]: Yes, you're right. I was trying to avoid too long and complicated names, but it seems unavoidable if we want to capture the meaning correctly.

File: src/gradualizer_lib.erl
Comment (erszcz) [reply src/gradualizer_lib.erl]: Updated. There's no distinction between minimum and fixed length binaries when printing, but there's a special case for empty binaries.

File: src/typechecker.erl
Comment (erszcz) [reply src/typechecker.erl]: Done ✅

**Comment by erszcz** [comment]
Unless you have more comments in response to today's changes, I think the only remaining thing is the wiki page. I'll create one once I find the next suitable time slot.

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:240]: Interesting, but why not simply `<<0:M>>`?

Comment (erszcz) [reply src/gradualizer_lib.erl:240]: Good point. This mislead me:

```
12> <<1:48>>.
<<0,0,0,0,0,1>>
```

But indeed, it's just:

```
13> merl:quote("<<1:48>>").
{bin,1,
     [{bin_element,1,{integer,1,1},{integer,1,48},default}]}
```

I'll simplify it.

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:232]: Good to simplify. Is there a need for `<<"aaaa...">>` or can we just use `<<0:M>>` here too? Is there is a nonempty string type in Elixir that would benefit from picking a string-looking value?

I think the minimal binary-picking would be just two clauses and no helper function needed, but it might have missed something.

```Erlang
pick_value(?type(binary, [{integer, _, M}, {integer, _, _}]), _Env) when M > 0 ->
    {bin, erl_anno:new(0), [{bin_element, 0, {integer, 0, 0}, {integer, 0, M}, default}]};
pick_value(?type(binary, _), _Env) ->
    {bin, erl_anno:new(0), []};
```

File: src/typechecker.erl
Comment (erszcz) [reply src/typechecker.erl:4446]: The latter actually leads to:

```
t2.erl: The variable on line 13 at column 11 is expected to have type non_neg_integer() but it has type <<_:_*8>>
```

but indeed, they don't lead to any exhaustiveness false positives, which is a good thing.

> since the real logic is in type_diff?

For the case when `should_bin_pat_enable_exhaustiveness_check()` returns true and the examples I could think of or you provide - yes. If `should_bin_pat_enable_exhaustiveness_check()` returns false we disable exhaustiveness checking for the clauses list completely, therefore avoid any false positives.

---

Ultimately, because of https://github.com/josefs/Gradualizer/pull/392#issuecomment-1145871376 we could just disable exhaustiveness checking for binary patterns completely (i.e. keep `refinable(binary()) == true`, but disable ex-check each time we see a binary pattern), and the goal to be able to include them in complex types would be met anyway. That would be as simple as possible, but the tradeoff would be that even the simplest binary patterns (for example an empty binary) would disable the ex-check on the complex type, too.

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [reply src/gradualizer_lib.erl]: Btw, I'd use `rem` in a guard, so the case clauses can be turned into function clauses, i.e. `when M > 0, M rem 8 =:= 0`.

Comment (zuiderkwast) [reply src/gradualizer_lib.erl]: Oh, you removed this. :-) Well, I'll leave it to you to decide if you want the `<<"aaa...">>` back or not. I don't mind, just want to understand the use cases.

File: src/typechecker.erl
Comment (zuiderkwast) [reply src/typechecker.erl:4446]: Thanks for the explanation.

Btw, I think my g/2 example could easily be corrected, e.g. `<<Y:1/binary, _/binary>>`.

Comment (zuiderkwast) [reply src/typechecker.erl:4446]: If you leave out the pattern `<<>>` in the examples `f/2` and the corrected `g/2`, you'd get "Values not covered: `<<>>`" and when you add a clause for `<<>>`, the error dissapears, which can make the user believe all values are covered. Hopefully, they have RTFM and found that exhaustiveness-checking is best effort (i.e. unsound) anyway, so all is fine. :-)

**Review by zuiderkwast — APPROVED** [review]
Feel free to merge when you have decided which variant of pick value you prefer.

**Comment by zuiderkwast** [comment]
Nice! But I miss the IMO most important info about the guarantees/soundness.

**Pull request merged**
